src/coord/src/coord.rs

```diff
@@ -867,6 +903,93 @@ impl<S: Append + 'static> Coordinator<S> {
             .unwrap();
     }
 
+    /// Attempts to commit all pending write transactions in a group commit. If the timestamp
+    /// chosen for the writes is not ahead of `now`, then we can execute and commit the writes
+    /// immediately. Otherwise we must wait for `now` to advance past the timestamp chosen for the
+    /// writes.
+    async fn try_group_commit(&mut self, timestamp: Option<Timestamp>) {
+        if self.pending_writes.is_empty() {
+            return;
+        }
+
+        let timestamp = if let Some(timestamp) = timestamp {
+            timestamp
+        } else {
+            self.peek_local_ts()
+        };
+        let now = (self.catalog.config().now)();
+
+        if timestamp > now {
+            // Cap retry time to 1s. In cases where the system clock has retreated by
+            // some large amount of time, this prevents against then waiting for that
+            // large amount of time in case the system clock then advances back to near
+            // what it was.
+            let remaining_ms =
+                std::cmp::min(timestamp.saturating_sub(now).saturating_add(1), 1_000);
+            let internal_cmd_tx = self.internal_cmd_tx.clone();
+            task::spawn(|| "group_commit", async move {
+                tokio::time::sleep(Duration::from_millis(remaining_ms)).await;
+                internal_cmd_tx
+                    .send(Message::GroupCommit(Some(timestamp)))
+                    .expect("sending to internal_cmd_tx cannot fail");
+            });
+        } else {
+            self.group_commit().await;
+        }
+    }
+
+    /// Commits all pending write transactions at the same timestamp. All pending writes will be
+    /// combined into a single Append command and sent to STORAGE as a single batch. All writes will
+    /// happen at the same timestamp and all involved tables will be advanced to some timestamp
+    /// larger than the timestamp of the write.
+    ///
+    /// If the table that some write was targeting has been deleted while the write was waiting,
+    /// then the write will be ignored and we respond to the client that the write was successful.
+    /// This is only possible if the write and the delete were concurrent. Therefore, we are free to
+    /// order the write before the delete without violating any consistency guarantees.
+    async fn group_commit(&mut self) {
+        // We cannot use the `peek_local_ts` from `try_group_commit`, because it may be invalid.
+        // While group commit was waiting, some DDL may have executed at the peeked timestamp
+        // causing that timestamp to close and become invalid for future writes. Therefore, we must
+        // get a new valid timestamp. This timestamp may be larger than the peeked timestamp and
+        // may even be larger than `now` depending on how much DDL was executed. In the future we
+        // should include DDL in group commits to avoid this issue.
```

Comment:
This should describe why it's ok to use a maybe future timestamp. Maybe:

> If a write timestamp is produced that is larger than `now`, the next batch of writes may trigger the wait loop in `try_group_commit`, so we will still prevent unboundedly advancing table timestamps ahead of the system clock.

Reply:
I moved this comment and added something that explains this (if I understood what you were saying).